crates/oidc-client/src/requests/jose.rs

```diff
@@ -0,0 +1,216 @@
+
+    jwt.verify_with_jwks(jwks)
+        .or(Err(JwtVerificationError::JwtSignatureNoKeyWorked))?;
+
+    let (header, mut claims) = jwt.clone().into_parts();
+
+    // Must have the proper issuer.
+    let iss = claims::ISS.extract_required(&mut claims)?;
+    if iss != issuer.as_str() {
+        return Err(JwtVerificationError::WrongIssuer);
+    }
+
+    // Must have the proper audience.
+    let aud = claims::AUD.extract_required(&mut claims)?;
+    if !aud.contains(client_id) {
+        return Err(JwtVerificationError::WrongAudience);
+    }
+
+    // Must use the proper algorithm.
+    if header.alg() != signing_algorithm {
+        return Err(JwtVerificationError::WrongSignatureAlg);
+    }
+
+    Ok(jwt.into_owned())
+}
+
+/// Decode and verify an ID Token.
+///
+/// Besides the checks of [`verify_signed_jwt()`], the following checks are
+/// performed:
+///
+/// * The `exp` claim must be present and the token must not have expired.
+///
+/// * The `iat` claim must be present must be in the past.
+///
+/// * The `sub` claim must be present.
+///
+/// If an authorization ID token is provided, these extra checks are performed:
+///
+/// * The `sub` claims must match.
+///
+/// * The `auth_time` claims must match.
+///
+/// # Arguments
+///
+/// * `id_token` - The serialized ID Token to decode and verify.
+///
+/// * `verification_data` - The data necessary to verify the ID Token.
+///
+/// * `auth_id_token` - If the ID Token is not verified during an authorization
+///   request, the ID token that was returned from the latest authorization
+///   request.
+///
+/// # Errors
+///
+/// Returns an error if the data is invalid or verification fails.
+pub fn verify_id_token(
+    id_token: &str,
+    verification_data: JwtVerificationData<'_>,
+    auth_id_token: Option<&Jwt<'_, HashMap<String, Value>>>,
+) -> Result<IdToken, IdTokenError> {
+    let id_token = verify_signed_jwt(id_token, verification_data)?;
+
+    let mut claims = id_token.payload().clone();
+
+    // Must not have expired.
+    claims::EXP.extract_required(&mut claims)?;
+
+    // `iat` claim must be present.
+    claims::IAT.extract_required(&mut claims)?;
+
+    // Subject identifier must be present.
+    let sub = claims::SUB.extract_required(&mut claims)?;
+
+    // No more checks if there is no previous ID token.
+    let auth_id_token = match auth_id_token {
+        Some(id_token) => id_token,
+        None => return Ok(id_token),
+    };
+
+    let mut auth_claims = auth_id_token.payload().clone();
+
+    // Subject identifier must always be the same.
+    let auth_sub = claims::SUB.extract_required(&mut auth_claims)?;
+    if sub != auth_sub {
+        return Err(IdTokenError::WrongSubjectIdentifier);
+    }
+
+    // If the authentication time is present, it must be unchanged.
+    if let Some(auth_time) = claims::AUTH_TIME.extract_optional(&mut claims)? {
+        let prev_auth_time = claims::AUTH_TIME.extract_required(&mut auth_claims)?;
```

Comment:
That's why the docs say that the ID token should be from the latest authentication request.

That's a validation step from OpenID Connect, and it shouldn't happen too often because this is for the refresh token endpoint, where the ID token is not required, and the `auth_time` claim should only be included in an ID token if it was explicity requested.